### PR TITLE
Add support for CONST monomial to Bounds system

### PR DESCRIPTION
**Pull request by GiudGiud**
refs #23448

### CODE REVIEW

File: framework/src/auxkernels/BoundsAuxBase.C [hunk 1]
```diff
@@ -33,23 +34,48 @@ BoundsAuxBase::BoundsAuxBase(const InputParameters & parameters)
...
+    paramError("bounded_variable", "Bounded variable must be nodal or of a CONSTANT order!");
+
+  auto & dummy =
+      _aux_sys.getActualFieldVariable<Real>(_tid, parameters.get<AuxVariableName>("variable"));
+
+  // Check that the dummy variable matches the bounded variable
+  if (dummy.feType() != _var.feType())
+    paramError(
+        "variable",
+        "Dummy bound aux. variable and bounded variable must both finite element order and family");
```
Comment (hugary1995) [inline framework/src/auxkernels/BoundsAuxBase.C:52]:
```suggestion
        "Dummy bound aux. variable and bounded variable must use the same finite element order and family");
```

File: framework/src/auxkernels/BoundsAuxBase.C [hunk 2]
```diff
@@ -33,23 +34,48 @@ BoundsAuxBase::BoundsAuxBase(const InputParameters & parameters)
...
+    if (_current_elem->n_dofs(_nl_sys.number(), _bounded_var.number()) > 0)
+    {
+      // The zero is for the component, this will only work for CONSTANT variables
+      dof_id_type dof = _current_elem->dof_number(_nl_sys.number(), _bounded_var.number(), 0);
 
-    Real bound = getBound();
-    _bounded_vector.set(dof, bound);
+      Real bound = getBound();
+      _bounded_vector.set(dof, bound);
+    }
```
Comment (hugary1995) [inline framework/src/auxkernels/BoundsAuxBase.C:78]: Write a separate function which returns the current dof number of the bounded variable, then remove the code duplication here.

File: test/tests/auxkernels/bounds/constant_bounds_elem.i [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    bound_type = lower
+    bound_value = -1
+  []
+[]
+
+[Executioner]
+  type = Steady
+
+  solve_type = 'PJFNK'
+[]
```
Comment (hugary1995) [inline test/tests/auxkernels/bounds/constant_bounds_elem.i:138]: IIRC you have to use `-snes_type vinewtonrsls` or `-snes_type vinewtonssls` for the bounds to be active.

Comment (hugary1995) [reply test/tests/auxkernels/bounds/constant_bounds_elem.i:138]: Can you also post of a plot of the result in this PR? Thanks.

**Review by hugary1995 — CHANGES_REQUESTED** [review]
Looks good overall. Thanks again for taking care of this. Useful stuff.

File: test/tests/auxkernels/bounds/constant_bounds_elem.i (hunk 3 above)
Comment (GiudGiud) [reply test/tests/auxkernels/bounds/constant_bounds_elem.i:138]: It's passed in the tests spec. It s a good point though I ll move them

Comment (GiudGiud) [reply test/tests/auxkernels/bounds/constant_bounds_elem.i:138]: Screen Shot 2023-02-16 at 3 36 47 PM

File: framework/src/auxkernels/BoundsAuxBase.C (hunk 2 above)
Comment (GiudGiud) [reply framework/src/auxkernels/BoundsAuxBase.C]: done. Only think was that this handled the "no local dof" case too. 

I think this matters if the block restriction is different between the dummy and the bounded

**Comment by GiudGiud** [comment]
Should be good now

File: framework/include/auxkernels/BoundsAuxBase.h [hunk 4]
```diff
@@ -49,4 +49,8 @@ class BoundsAuxBase : public AuxKernel
 
   /// MOOSE variable we set the bound for
   MooseVariable & _var;
+
+private:
+  /// Return the current DOF index to apply the bound on
+  dof_id_type getDoFIndex();
```
Comment (lindsayad) [inline framework/include/auxkernels/BoundsAuxBase.h:55]:
```suggestion
  dof_id_type getDoFIndex() const;
```

File: framework/src/auxkernels/BoundsAuxBase.C [hunk 5]
```diff
@@ -33,24 +34,59 @@ BoundsAuxBase::BoundsAuxBase(const InputParameters & parameters)
...
-    mooseError("BoundsAuxBase must be used on a nodal auxiliary variable!");
   if (!Moose::PetscSupport::isSNESVI(*dynamic_cast<FEProblemBase *>(&_subproblem)))
     mooseDoOnce(mooseWarning(
         "A variational inequalities solver must be used in conjunction with BoundsAux"));
+
+  // Check that the bounded variable is of a supported type
+  if (!_var.isNodal() && (_var.feType().order != CONSTANT))
+    paramError("bounded_variable", "Bounded variable must be nodal or of a CONSTANT order!");
+
+  auto & dummy =
```
Comment (lindsayad) [inline framework/src/auxkernels/BoundsAuxBase.C:45]:
```suggestion
  const auto & dummy =
```

File: framework/src/auxkernels/BoundsAuxBase.C [hunk 6]
```diff
@@ -33,24 +34,59 @@ BoundsAuxBase::BoundsAuxBase(const InputParameters & parameters)
...
+      _aux_sys.getActualFieldVariable<Real>(_tid, parameters.get<AuxVariableName>("variable"));
+
+  // Check that the dummy variable matches the bounded variable
+  if (dummy.feType() != _var.feType())
+    paramError("variable",
+               "Dummy bound aux. variable and bounded variable must use the same finite element "
```
Comment (lindsayad) [inline framework/src/auxkernels/BoundsAuxBase.C:51]:
```suggestion
               "Dummy bounds aux variable and bounded variable must use the same finite element "
```

File: test/tests/auxkernels/bounds/constant_bounds_elem.i [hunk 7]
```diff
@@ -0,0 +1,144 @@
+[Mesh]
+  type = GeneratedMesh
+
+  dim = 1
+
+  xmin = 0
+  xmax = 1
+
+  nx = 10
```
Comment (lindsayad) [inline test/tests/auxkernels/bounds/constant_bounds_elem.i:9]:
```suggestion
  type = GeneratedMesh
  dim = 1
  xmin = 0
  xmax = 1
  nx = 10
```

File: framework/src/auxkernels/BoundsAuxBase.C [hunk 8]
```diff
@@ -33,24 +34,59 @@ BoundsAuxBase::BoundsAuxBase(const InputParameters & parameters)
...
+  {
     Real bound = getBound();
     _bounded_vector.set(dof, bound);
   }
 
   return 0.0;
 }
+
+dof_id_type
+BoundsAuxBase::getDoFIndex()
```
Comment (lindsayad) [inline framework/src/auxkernels/BoundsAuxBase.C:70]:
```suggestion
BoundsAuxBase::getDoFIndex() const
```

File: test/tests/auxkernels/bounds/constant_bounds_elem.i [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+[]
+
+[Variables]
+  [u]
+    order = CONSTANT
+    family = MONOMIAL
+  []
+
```
Comment (lindsayad) [inline test/tests/auxkernels/bounds/constant_bounds_elem.i:17]:
```suggestion
```

File: test/tests/auxkernels/bounds/constant_bounds_elem.i [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+[Kernels]
+  [diff_u]
+    type = Diffusion
+    variable = u
+  []
+  [reaction_u]
+    type = Reaction
+    variable = u
+  []
+
```
Comment (lindsayad) [inline test/tests/auxkernels/bounds/constant_bounds_elem.i:40]:
```suggestion
```

File: test/tests/auxkernels/bounds/constant_bounds_elem.i [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+[BCs]
+  [left_u]
+    type = DGFunctionDiffusionDirichletBC
+    variable = u
+    boundary = '0'
+    function = -0.5
+    epsilon = -1
+    sigma = 6
+  []
+
```
Comment (lindsayad) [inline test/tests/auxkernels/bounds/constant_bounds_elem.i:77]:
```suggestion
```

File: test/tests/auxkernels/bounds/constant_bounds_elem.i [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+  [right_u]
+    type = NeumannBC
+    variable = u
+    boundary = 1
+    value = 30
+  []
+
```
Comment (lindsayad) [inline test/tests/auxkernels/bounds/constant_bounds_elem.i:84]:
```suggestion
```

File: test/tests/auxkernels/bounds/constant_bounds_elem.i [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+  [left_v]
+    type = DGFunctionDiffusionDirichletBC
+    variable = v
+    boundary = '0'
+    function = 4
+    epsilon = -1
+    sigma = 6
+  []
+
```
Comment (lindsayad) [inline test/tests/auxkernels/bounds/constant_bounds_elem.i:93]:
```suggestion
```

File: test/tests/auxkernels/bounds/constant_bounds_elem.i [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+    bound_value = 1
+  []
+  [u_lower_bound]
+    type = ConstantBoundsAux
+    variable = bounds_dummy
+    bounded_variable = u
+    bound_type = lower
+    bound_value = 0
+  []
+
```
Comment (lindsayad) [inline test/tests/auxkernels/bounds/constant_bounds_elem.i:117]:
```suggestion
```

File: test/tests/auxkernels/bounds/constant_bounds_elem.i [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+    variable = bounds_dummy
+    bounded_variable = v
+    bound_type = lower
+    bound_value = -1
+  []
+[]
+
+[Executioner]
+  type = Steady
+
```
Comment (lindsayad) [inline test/tests/auxkernels/bounds/constant_bounds_elem.i:136]:
```suggestion
```

File: framework/src/auxkernels/BoundsAuxBase.C [hunk 16]
```diff
@@ -33,24 +34,59 @@ BoundsAuxBase::BoundsAuxBase(const InputParameters & parameters)
...
 Real
 BoundsAuxBase::computeValue()
 {
-  if (_current_node->n_dofs(_nl_sys.number(), _bounded_var.number()) > 0)
-  {
-    // The zero is for the component, this will only work for Lagrange variables!
-    dof_id_type dof = _current_node->dof_number(_nl_sys.number(), _bounded_var.number(), 0);
+  dof_id_type dof = getDoFIndex();
 
+  if (dof != std::numeric_limits<dof_id_type>::max())
```
Comment (hugary1995) [inline framework/src/auxkernels/BoundsAuxBase.C:60]: I think this condition will never be true based on the logic below. So maybe change this to an assertion.

File: test/tests/auxkernels/bounds/tests [hunk 17]
```diff
@@ -1,31 +1,57 @@
...
-  [../]
+  []
+  [constant_bounds_elem]
+    type = 'Exodiff'
+    input = 'constant_bounds_elem.i'
+    exodiff = 'constant_bounds_elem_out.e'
+    scale_refine = 3
+    design = 'ConstantBoundsAux.md'
+    issues = '#951 #14946'
+    requirement = "The system shall include the ability to compute an constant monomial field variable with a constant upper and lower bounds."
```
Comment (lindsayad) [inline test/tests/auxkernels/bounds/tests:27]:
```suggestion
    requirement = "The system shall include the ability to compute a constant monomial field variable with constant upper and lower bounds."
```

**Review by hugary1995 — DISMISSED** [review]

File: test/tests/auxkernels/bounds/tests [hunk 18]
```diff
@@ -1,31 +1,57 @@
...
+  []
+
+  [exceptions]
+    requirement = "The system shall "
```
Comment (lindsayad) [inline test/tests/auxkernels/bounds/tests:31]:
```suggestion
    requirement = "The system shall"
```
Details come out as bullets so space not necessary

File: test/tests/auxkernels/bounds/tests [hunk 19]
```diff
@@ -1,31 +1,57 @@
...
+    issues = '#14946'
+    design = 'ConstantBoundsAux.md'
+    [solver]
+      type = 'RunApp'
+      input = 'constant_bounds.i'
+      expect_out = "A variational inequalities solver must be used in conjunction with BoundsAux"
+      allow_warnings = true
+      cli_args = "Executioner/petsc_options_iname='' Executioner/petsc_options_value=''"
+      detail = 'emit a warning if a variational inequalities (VI) solver is not found when BoundsAux is provided.'
```
Comment (lindsayad) [inline test/tests/auxkernels/bounds/tests:40]:
```suggestion
      detail = 'emit a warning if a variational inequalities (VI) solver is not found when the bounds system is used.'
```

**Review by lindsayad — CHANGES_REQUESTED** [review]

File: framework/src/auxkernels/BoundsAuxBase.C (hunk 16 above)
Comment (GiudGiud) [reply framework/src/auxkernels/BoundsAuxBase.C:60]: I think it s a block restriction safeguard. This preserves the old behavior

Comment (lindsayad) [reply framework/src/auxkernels/BoundsAuxBase.C:60]: @hugary1995 it is possible to be true. Think of a first order Lagrange variable on a second order mesh. There will be no dofs on non-vertex nodes

**Comment by GiudGiud** [comment]
should be good?

File: framework/src/auxkernels/BoundsAuxBase.C (hunk 16 above)
Comment (hugary1995) [reply framework/src/auxkernels/BoundsAuxBase.C:60]: I see. @GiudGiud can you add the reasoning as comments in the source code?

Comment (GiudGiud) [reply framework/src/auxkernels/BoundsAuxBase.C:60]: sure

**Comment by GiudGiud** [comment]
I think this PR has pretty strong endorsement at this point, down for a review

**Review by lindsayad — APPROVED** [review]

**Comment by lindsayad** [comment]
Did someone circle a petition? 😆

**Pull request merged**